common/archiver/README.md

```diff
@@ -0,0 +1,69 @@
+```
+2. Each Archiver will be given a `BootstrapContainer` when it's created and the archiver implementation can (will) use them to read history, emit logs/metrics etc. Check `interface.go` for more details.
+
+3. Implement the Archiver interface. Both Archiver interface contains three methods:
+  
+```go
+func Archive(ctx context.Context, URI string, request *ArchiveRequest, opts ...ArchiveOption) error
+
+func Get(ctx context.Context, URI string, request *GetRequest) (*GetResponse, error)
+
+func ValidateURI(URI string) error
+```
+
+Each archiver implementation should define a format of URI, which is used to describe the type of implementation and the target location of `Archive()` or `Get()` operation. For example, the URI for filestore uses the format: `file:///path/to/dir`. The `ValidateURI()` method is responsible for checking if a given string is a valid URI.
+
+The `Archive()` method should archive workflow histories or visibility records described by the archive request to the location specified by the URI, while the `Get()` method is responsible for retrieving those archived data.
+
+Since the `Archive()` method will be called from different places (for example, timerQueueProcessor and worker), it may want to do different things. Therefore, a list of `ArchiveOption` will be passed to this method. These options will be applied to an `ArchiveFeatureCatalog`, and by checking the fields of that catalog, the `Archive()` method can figure out what to do. Right now, the only feature in the catalog is `ProgressManager` which can be used to record and load archive progress. More features can be added if needed. The correct way to use it is shown in the code sample below.
+
+```go
+func (a *Archiver) Archive(
+	ctx context.Context,
+	URI string,
+	request *ArchiveRequest,
+	opts ...ArchiveOption,
+) error {
+  featureCatalog := **GetFeatureCatalog**(opts...) // this function is defined in options.go
+
+  var progress progress
+
+  // Check if the feature for recording progress is enabled.
+  if featureCatalog.ProgressManager != nil {
+    if err := featureCatalog.ProgressManager.LoadProgress(ctx, &prevProgress); err != nil {
+      // log some error message and return error if needed.
+    }
+  }
+
+  // Your archiver implementation...
+
+  // Record current progress
+  if featureCatalog.ProgressManager != nil {
+    if err := featureCatalog.ProgressManager.RecordProgress(ctx, progress); err != nil {
+      // log some error message and return error if needed. 
+    }
+  }
+}
+```
+
+4. The `archiver` package provides a utility class called `HistoryIterator` which is a wrapper of `HistoryReader`. Its usage is simpler than the `HistoryReader` given in the `BootstrapContainer`, so archiver implementations can choose to use it when reading workflow histories. Check the `historyIterator.go` file for more details. Sample usage can be found in the filestore historyArchiver implementation.
```

Comment:
My understanding was that HistoryReader would go away and we only needed HistoryIterator now that we can reconstruct its state?

Reply:
Do you mean we should provide HistoryIterator instead of HistoryReader in the BootstrapParams? And then export the reset() method of HistoryIterator so that it can be called inside the Archive() method? 

But currently implementation of HistoryIterator is linked to a particular archive request. If that problem is solved, an other problem we need to address is that: the config for the historyIterator is stored in the provider, so the caller of the GetHistoryArchiver() should give the provider a historyReader and historyIterator will be created inside the provider. Or we store the config to history/worker service instead of provider. Both method seems weird to me.

Reply:
My bad. It should be HistoryManager, not HistoryReader